File: LeetCode/Medium/0005-longest-palindromic-substring/0005-longest-palindromic-substring.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        int n = s.size();
        vector<vector<bool>> dp(n, vector<bool>(n, false));
        for(int i=0; i<n; i++)
            dp[i][i] = true;
        string ans = string(1, s[n-1]);

        int maxLen = 1;
        for(int lo = n-1; lo > -1; lo--){
            for(int hi = lo+1; hi < n; hi++){
                if(s[lo] == s[hi]){
                    if(hi - lo == 1 or dp[lo+1][hi-1]){
                        dp[lo][hi] = true;
                        if(maxLen < hi - lo + 1){
                            maxLen = hi - lo + 1;
                            ans = s.substr(lo, maxLen);
                        }
                    }
                }
            }
        }
        return ans;
    }
};

File: LeetCode/Medium/0005-longest-palindromic-substring/0005-longest-palindromic-substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        n = len(s)
        dp = [[False] * n for _ in range(n)]

        ans = ''
        for i in range(n):
            dp[i][i] = True
        ans = s[-1]

        maxLen = 1
        for lo in range(n-1, -1, -1):
            for hi in range(lo+1, n):
                if s[lo] == s[hi]:
                    if hi - lo == 1 or dp[lo+1][hi-1]:
                        dp[lo][hi] = True
                        # print(lo, hi)
                        if maxLen < hi - lo + 1:
                            maxLen = hi - lo + 1
                            ans = s[lo : hi+1]
        return ans